# Handling Imbalanced Dataset

In [1]:
# Importing the libraries
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('credit_data.csv')

In [3]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

### Class
- 0  ->  284315  -> Legit Transactions
- 1  ->     492  -> Fraudulent Transactions
- This is the Highly imbalanced dataset

In [7]:
legit = df[df['Class'] == 0]
fraud = df[df['Class'] == 1]

In [8]:
print(legit.shape, fraud.shape)

(284315, 31) (492, 31)


## Under-Sampling
Build a sample dataset containing equal number of distribution of classes Legit and Fraud.
- No of Fraud Transactions = 492

In [9]:
legit_sample = legit.sample(n=492)

In [10]:
legit_sample.shape

(492, 31)

In [12]:
# Concatenate two datasets
new_df = pd.concat([legit_sample, fraud], axis=0)

In [13]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
3618      3094.0 -0.612026  0.938788  1.354407  0.955059 -0.523281  0.191454   
250380  154874.0  1.746113  0.277682 -0.674812  3.818199  0.233324 -0.092876   
123396   76910.0  1.247000  0.628052  0.369489  2.284129  0.275468 -0.038581   
173260  121410.0  2.093305 -0.841304 -0.372266 -0.361620 -0.960115 -0.128898   
221438  142570.0  1.827777  0.267640 -0.148190  3.782153  0.170809  0.653658   

              V7        V8        V9  ...       V21       V22       V23  \
3618    1.018194 -0.042082 -0.317349  ... -0.146281 -0.411536  0.209932   
250380  0.259947  0.015660 -1.031846  ... -0.149639 -0.625938  0.255574   
123396  0.158325 -0.105334 -0.785854  ... -0.118596 -0.346265 -0.090721   
173260 -1.028501  0.080810 -0.279499  ... -0.496801 -0.926750  0.417371   
221438 -0.263134  0.201327 -0.373678  ... -0.411304 -1.245874  0.506208   

             V24       V25       V26       V27       V28  Amount  Class  
3618    0.005353 -0.776464 -0.667928  0.019674  0.116949  130.20      0  
250380 -0.018957 -0.259852 -0.292829 -0.063101 -0.046457   72.63      0  
123396 -0.449027  0.559450 -0.024928 -0.005078  0.017939    4.72      0  
173260 -0.564764 -0.674912  0.005952  0.006590 -0.047912    8.00      0  
221438  0.481327 -0.566654 -0.533869 -0.009424 -0.019718   33.51      0  

[5 rows x 31 columns]

In [15]:
new_df.shape

(984, 31)

In [17]:
new_df['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64